packages/dds/tree/src/test/simple-tree/arrayNode.spec.ts

```diff
@@ -0,0 +1,62 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { strict as assert } from "assert";
+
+import {
+	ImplicitAllowedTypes,
+	InsertableTreeFieldFromImplicitField,
+	InsertableTreeNodeFromImplicitAllowedTypes,
+	NodeKind,
+	SchemaFactory,
+	TreeArrayNode,
+	TreeFieldFromImplicitField,
+	TreeNodeSchema,
+} from "../../simple-tree/index.js";
+
+import { hydrate } from "./utils.js";
+import { Mutable } from "../../util/index.js";
+
+type ArrayNodeSchema<T extends ImplicitAllowedTypes> = TreeNodeSchema<
+	string,
+	NodeKind.Array,
+	TreeArrayNode<T>,
+	Iterable<InsertableTreeNodeFromImplicitAllowedTypes<T>>,
+	true
+>;
+
+const schemaFactory = new SchemaFactory("Test");
+
+function hydrateArray<T extends TreeNodeSchema | readonly TreeNodeSchema[]>(
+	schemaType: "structural" | "class-based",
```

Comment:
I'm not sure if this is a good distinction. is this trying to select between a structurally named schema and an explicitly named one, or "Customizable" and "POJO Emulation" modes (see docs on SchemaFactory)? It happens to be control both of those since our API couples them, but I don't think the intention or actual effects are particularly clear from the strings you chose.

Reply:
I'm just trying to test the two flavors of the API that users can access and make sure that they do what we think they do. Right now the docs for `array` say "structurally typed" and "class" depending on which version you pick, so I think the strings here are aligned well.

Reply:
It happens to be the case that "structurally typed" arrays are unconditionally in pojo-emulation mode (since customizing them wouldn't make sense) but is not technically related to them being named structurally and is a separate thing. The fact that other kinds of array schema never are in pojo-emulation mode is an arbitrary choice at the schema factory layer (which this is not supposed to depend on!), and not consistent with how objects schema work so might be something we change.

Your tests are not testing the structural typing nature of the array's identifier: you are testing how non-customizable arrays work and "class-based" is not a description that we used anywhere that I'm aware of.

I still think explicitly naming these based on the thing you are actually testing would be better, and removing the complex helper/test wrapper function and doing everything how actual normal users would do it woul dbe much more readable, ex:

```typescript
class CustomizableArray extends factory.array("Customizable", factory.number) {}
const PojoEmulationArray = factory.array(factory.number);
type PojoEmulationArray = NodeFromSchema<typeof PojoEmulationArray>;
for (const [type, schemaType] of [
	["Customizable", CustomizableArray],
	["PojoEmulation", PojoEmulationArray],
] as const) {
	describe(type, () => {
		it("fails at runtime if attempting to set content via index assignment", () => {
			const array = hydrate(schemaType, [0]);
			const mutableArray = array as Mutable<typeof array>;
			assert.equal(mutableArray.length, 1);
			assert.throws(() => (mutableArray[0] = 3)); // An index within the array that already has an element
			assert.throws(() => (mutableArray[1] = 3)); // An index just past the end of the array, where a new element would be pushed
			assert.throws(() => (mutableArray[2] = 3)); // An index that would leave a "gap" past the current end of the array if a set occurred
		});

		it("stringifies in the same way as a JS array", () => {
			const jsArray = [0, 1, 2];
			const array = hydrate(schemaType, jsArray);
			assert.equal(JSON.stringify(array), JSON.stringify(jsArray));
		});
	});
}
```

Reply:
Thanks for the example code. I decided to declare the two kids of array schema at the top, as you did, rather than having the function that creates them. The downside is that they are hardcoded to `number` now, but I suppose if we want to have tests that do stuff with other data types then we can just make it hardcoded to the union.

For readability and simplicity, I prefer having "two plain-ol describe blocks and a helper function" over a loop that runs describe from tuples (and I was able to get to be quite simple after removing the `hydrateArray` function).

As far as the naming, I remain unconvinced - in fact, you've convinced me further that we should keep the name as-is. My interest here is the following: 
* If the user makes an array via the `sf.array(contentTypes)` function, what is the behavior of the resulting object? 
* If the user makes an array via the `sf.array("...", contentTypes)` function, what is the behavior of the resulting object?

In the public function documentation, and colloquially amongst our team, we call the former thing a "structurally typed" array. In the public function documentation, and colloquially amongst our team, we call the latter thing a "class". That's a pretty clear match to the current names.

Suppose I named these tests "pojo emulation mode" and "other thing". Then, as you pointed out above, if we were to change the policy of when the schema factory layer employs pojo-emulation mode and when it does not, then these tests would no longer be accurate and would no longer be testing what they say they are testing. With the current names, they would remain accurate. I think the conclusion here is that **these tests are technically not testing pojo-emulation mode, they are testing the two flavors of the public API**. I'm fine with that right now. I am not considering that policy changing to be a legitimate concern in the near future. If it does become a possibility, then we should change all these tests to directly construct the objects at a lower lever rather than going through the public API. In the meantime, I think it's fine to use our public API to keep things simple.

Reply:
refer to the table in the doc comment for SchemaFactory for details on POJO emulation mode vs Customizable. Align to the terminology used there. See also https://github.com/microsoft/FluidFramework/pull/20674/files#diff-ff508c06cf82b9008e19b5ab48fa6eeb14bc839dba85b7194d06328b42f3dfa9R54 .

> these tests are technically not testing pojo-emulation mode, they are testing the two flavors of the public API
These prabably should be unit tests for the functionality implemented in arrayNode.ts, which does not include the public API entry point, and also does not contain the structural naming code, or the policy decision to use POFO mode only when structurally named. It would be possible to allow non structurally named POJO mode arrays without modifying arrayNode.ts so it seems really odd to tie testing of the logic in this file, which has no control over naming, based on how its named, not what its implementing.